File: python/nvagentrt/llm.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""LLM lifecycle operations.

Provides both manual and managed LLM-call workflows, including streaming.

Functions:
    call(name, request, *, handle=None, attributes=None, data=None, metadata=None, model_name=None)
        Begin an LLM call manually. Returns an ``LLMHandle`` that must later
        be passed to ``call_end``. Emits a ``Start`` event. The optional
        ``model_name`` identifies the LLM model and is propagated to events
        for ATIF trajectory export.

    call_end(handle, response, *, data=None, metadata=None)
        End a manual LLM call. Records the response and emits an ``End`` event.

    execute(name, request, func, *, handle=None, attributes=None, data=None, metadata=None, model_name=None)
        Execute an LLM call through the full middleware pipeline (request
        intercepts -> sanitize-request guardrails -> conditional-execution
        guardrails -> execution intercepts -> *func* -> response intercepts ->
        sanitize-response guardrails). Returns an awaitable of the final response.
        The optional ``model_name`` is propagated to events for ATIF trajectory export.

    stream_execute(name, request, func, collector, finalizer, *, handle=None,
            attributes=None, data=None, metadata=None, model_name=None)
        Like ``execute`` but the execution function returns an async iterator
        of SSE text chunks. The ``collector`` callable is invoked with each
        intercepted chunk (after stream response intercepts). The ``finalizer``
        callable is invoked once when the stream is exhausted and returns the
        aggregated response as a JSON-serializable value. Returns an awaitable
        ``LlmStream`` that can be iterated with ``async for``.
        Stream-response intercepts are applied to each SSE event in flight.

    request_intercepts(request)
        Run the registered LLM request intercept chain on the given request.
        Returns the transformed ``LLMRequest``.

    conditional_execution(request)
        Run the registered LLM conditional execution guardrail chain.
        Raises ``RuntimeError`` if any guardrail rejects.

    response_intercepts(response)
        Run the registered LLM response intercept chain on the given response.
        Returns the transformed response.

Example::

    import nvagentrt

    req = nvagentrt.LLMRequest("POST", "https://api.example.com/chat", {}, body)

    # Non-streaming
    resp = await nvagentrt.llm.execute("gpt-4", req, my_llm_fn)

    # Streaming with collector/finalizer
    chunks = []
    def collect(chunk: str) -> None:
        chunks.append(chunk)
    def finalize() -> dict:
        return {"content": "".join(chunks)}

    stream = await nvagentrt.llm.stream_execute(
        "gpt-4", req, my_stream_fn, collect, finalize,
    )
    async for chunk in stream:
        print(chunk, end="")
"""

from nvagentrt._native import (
    nvagentrt_llm_call as call,
)
from nvagentrt._native import (
    nvagentrt_llm_call_end as call_end,
)
from nvagentrt._native import (
    nvagentrt_llm_call_execute as execute,
)
from nvagentrt._native import (
    nvagentrt_llm_conditional_execution as conditional_execution,
)
from nvagentrt._native import (
    nvagentrt_llm_request_intercepts as request_intercepts,
)
from nvagentrt._native import (
    nvagentrt_llm_response_intercepts as response_intercepts,
)
from nvagentrt._native import (
    nvagentrt_llm_stream_call_execute as stream_execute,
)

__all__ = [
    "call",
    "call_end",
    "execute",
    "stream_execute",
    "request_intercepts",
    "conditional_execution",
    "response_intercepts",
]

File: go/nvagentrt/llm/llm.go
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Package llm provides shorthand access to NVAgentRT LLM call operations.
//
// It re-exports the core LLM lifecycle functions (LlmCall, LlmCallEnd,
// LlmCallExecute, LlmStreamCallExecute) under shorter names for convenience.
//
// Example usage:
//
//	import "github.com/nvidia/nvagentrt/go/nvagentrt/llm"
//
//	native := map[string]interface{}{"model": "gpt-4", "messages": []interface{}{}}
//	result, err := llm.Execute("chat", native,
//	    func(nativeJSON json.RawMessage) (json.RawMessage, error) {
//	        // ... call the LLM API ...
//	        return json.RawMessage(`{"choices":[]}`), nil
//	    },
//	)
package llm

import (
	"encoding/json"

	"github.com/nvidia/nvagentrt/go/nvagentrt"
)

// Call starts an LLM call lifecycle and returns an [nvagentrt.LLMHandle],
// emitting a Start event. End the call with [CallEnd]. This is a shorthand for
// [nvagentrt.LlmCall].
func Call(name string, native interface{}, opts ...nvagentrt.LLMCallOption) (*nvagentrt.LLMHandle, error) {
	return nvagentrt.LlmCall(name, native, opts...)
}

// CallEnd completes an LLM call that was started with [Call], emitting an End
// event. This is a shorthand for [nvagentrt.LlmCallEnd].
func CallEnd(handle *nvagentrt.LLMHandle, response json.RawMessage, opts ...nvagentrt.LLMCallOption) error {
	return nvagentrt.LlmCallEnd(handle, response, opts...)
}

// Execute runs a complete LLM call lifecycle with the full middleware pipeline
// (conditional-execution guardrails, request intercepts, sanitize-request
// guardrails, execution intercepts, fn, response intercepts, sanitize-response
// guardrails) and returns the final response JSON. This is a shorthand for
// [nvagentrt.LlmCallExecute].
func Execute(name string, native interface{}, fn nvagentrt.LLMExecutionFunc, opts ...nvagentrt.LLMCallOption) (json.RawMessage, error) {
	return nvagentrt.LlmCallExecute(name, native, fn, opts...)
}

// StreamExecute runs a streaming LLM call lifecycle with the full middleware
// pipeline (conditional-execution guardrails run first on the raw request) and
// returns an [nvagentrt.LlmStream] for consuming JSON chunks. This is a
// shorthand for [nvagentrt.LlmStreamCallExecute].
//
// The collector callback is invoked with each intercepted chunk JSON for
// accumulation. The finalizer callback is invoked once when the stream is
// exhausted and must return a JSON string representing the aggregated response.
// Pass nil for either to use the default no-op behavior.
func StreamExecute(name string, native interface{}, fn nvagentrt.LLMExecutionFunc, collector nvagentrt.CollectorFunc, finalizer nvagentrt.FinalizerFunc, opts ...nvagentrt.LLMCallOption) (*nvagentrt.LlmStream, error) {
	return nvagentrt.LlmStreamCallExecute(name, native, fn, collector, finalizer, opts...)
}

// RequestIntercepts runs the registered LLM request intercept chain on the
// given request and returns the transformed request. This is a shorthand for
// [nvagentrt.LlmRequestIntercepts].
func RequestIntercepts(request json.RawMessage) (json.RawMessage, error) {
	return nvagentrt.LlmRequestIntercepts(request)
}

// ConditionalExecution runs the registered LLM conditional execution guardrail
// chain. Returns nil if all guardrails pass, or an error with the rejection
// reason if blocked. Optional [nvagentrt.LLMCallOption] values can supply a
// [nvagentrt.WithLLMToRequest] converter. This is a shorthand for
// [nvagentrt.LlmConditionalExecution].
func ConditionalExecution(request json.RawMessage, opts ...nvagentrt.LLMCallOption) error {
	return nvagentrt.LlmConditionalExecution(request, opts...)
}

// ResponseIntercepts runs the registered LLM response intercept chain on the
// given response and returns the transformed response. This is a shorthand for
// [nvagentrt.LlmResponseIntercepts].
func ResponseIntercepts(response json.RawMessage) (json.RawMessage, error) {
	return nvagentrt.LlmResponseIntercepts(response)
}
